### PR TITLE
Fix links in mails with hyphen

### LINKED ISSUE #300: Linkify creates http link in mailto link when host contains a hyphen
Linkify creates http links **in** mailto links if the mail host name contains a hyphen `-` (and `parse_email=True`). The http link starts with the hyphen, even if host names mustn't start with hyphens at all (https://en.wikipedia.org/wiki/Hostname#Restrictions_on_valid_hostnames).

```py
>>> linkify('foo@exa-mple.com', parse_email=True)
'<a href="mailto:foo@exa-mple.com">foo@exa<a href="http://-mple.com" rel="nofollow">-mple.com</a></a>'
```
This is what the link looks like: <a href="mailto:foo@exa-mple.com">foo@exa<a href="http://-mple.com" rel="nofollow">-mple.com</a></a> :smile:

### PR DESCRIPTION
**Pull request by sedrubal**
Do not create a http link inside a mailto link

Partly fixes #300:

```py
>>> linkify('foo@exa-mple.com', parse_email=True)
'<a href="mailto:foo@exa-mple.com">foo@exa<a href="http://-mple.com" rel="nofollow">-mple.com</a></a>'
'<a href="mailto:foo@exa-mple.com">foo@exa-mple.com</a>'
```

but it does not fix this issue:

```py
>>> linkify('foo@bar-baz.com')
'foo@bar<a href="http://-baz.com" rel="nofollow">-baz.com</a>'
```

### CODE REVIEW

**Comment by willkg** [comment]
I've been swamped. I'll try to get to this soon. Possibly early September.

**Comment by sedrubal** [comment]
did you find some time to look over this?

**Comment by willkg** [comment]
Not, yet. I've got a few Bleach things in my queue and some html5lib things. Trying to carve out a large block of time to work through them. Maybe next week.

Sorry for the slowness! :(

File: .gitignore [hunk 1]
```diff
@@ -10,3 +10,4 @@ build
 docs/_build/
 .cache/
 .eggs/
+.*env*/
```
Comment (willkg) [inline .gitignore]: Why did you add this?

File: bleach/linkifier.py [hunk 2]
```diff
@@ -349,7 +349,17 @@ def strip_non_url_bits(self, fragment):
...
         for token in src_iter:
+            if in_a:
+                if token['type'] == 'EndTag' and token['name'] == 'a':
+                    in_a = False
+                yield token
+                continue
+            elif token['type'] == 'StartTag' and token['name'] == 'a':
+                in_a = True
+                yield token
+                continue
```
Comment (willkg) [inline bleach/linkifier.py]: There's a bunch of "are we in an `<a>`" code and bookkeeping in this class. I think maybe we should adjust those cases rather than add another bookkeeping measure.

Namely, why is this method getting called if we're in an `<a>` tag at all?

Comment (willkg) [reply bleach/linkifier.py]: My apologies--I see what's going on now.

**Comment by willkg** [comment]
Thank you!

File: .gitignore (hunk 1 above)
Comment (sedrubal) [reply .gitignore]: I'm sorry for adding this change to this pull request. It's for ignoring virtual environments.

**Comment by sedrubal** [comment]
Thanks for merging.
